### PR TITLE
feat(optimistic-oracle-client): Can read SkinnyOptimisticOracle state

### PR DESCRIPTION
**Pull request by nicholaspai**
**Motivation**

This is a prerequisite PR to enabling the `optimistic-oracle-proposer` to work with the `SkinnyOptimisticOracle` contract. This should make the OptimisticOracleClient work with all existing bots without modifications, but also enable the caller to specify the client to detect a SkinnyOptimisticOracle.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts [hunk 1]
```diff
@@ -144,33 +195,31 @@ export class OptimisticOracleClient {
       return hasDispute === undefined;
     });
 
-    const unsettledProposals = (
```
Comment (nicholaspai) [inline packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts:147]: This logic was actually uneccessary as we can simple filter out any events that overlap with `Settle` events.

File: packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts [hunk 2]
```diff
@@ -212,36 +261,28 @@ export class OptimisticOracleClient {
     ).filter(isDefined);
 
     // Filter out disputes that were already settled and reformat data.
-    const unsettledResolvedDisputeEvents = (
```
Comment (nicholaspai) [inline packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts:215]: Same reasoning as [this comment](https://github.com/UMAprotocol/protocol/pull/3554/files#r742089288)

File: packages/financial-templates-lib/test/clients/OptimisticOracleClient.js [hunk 3]
```diff
@@ -126,203 +127,418 @@ describe("OptimisticOracleClient.js", function () {
     );
   });
 
-  it("Basic proposal lifecycle: request, propose, expire without dispute", async function () {
```
Comment (nicholaspai) [inline packages/financial-templates-lib/test/clients/OptimisticOracleClient.js:129]: These tests are moved unmodified into a `describe()` block to denote that its working with the default `OptimisticOracle` contract.

File: packages/financial-templates-lib/test/clients/OptimisticOracleClient.js [hunk 4]
```diff
@@ -126,203 +127,418 @@ describe("OptimisticOracleClient.js", function () {
...
+      const currentTime = Number((await web3.eth.getBlock("latest")).timestamp);
+      await advanceBlockAndSetTime(web3, currentTime + 1);
+      await advanceBlockAndSetTime(web3, currentTime + 2);
+
+      await clientShortLookback.update();
+      result = clientShortLookback.getUnproposedPriceRequests();
+      objectsInArrayInclude(result, []);
+    });
   });
+  describe("Skinny OptimisticOracle contract", function () {
```
Comment (nicholaspai) [inline packages/financial-templates-lib/test/clients/OptimisticOracleClient.js:331]: These tests are new and follow the exact same flow as in the traditional OptimisticOracle tests

File: packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts [hunk 5]
```diff
@@ -1,36 +1,68 @@
...
   OptimisticOracleWeb3,
   VotingAncillaryInterfaceTestingWeb3,
   OptimisticOracleWeb3Events,
+  SkinnyOptimisticOracleWeb3Events,
 } from "@uma/contracts-node";
 
-type RequestPrice = OptimisticOracleWeb3Events.RequestPrice["returnValues"];
-type ProposePrice = OptimisticOracleWeb3Events.ProposePrice["returnValues"];
-type DisputePrice = OptimisticOracleWeb3Events.DisputePrice["returnValues"];
+type SkinnyRequestPrice = SkinnyOptimisticOracleWeb3Events.RequestPrice;
```
Comment (chrismaree) [inline packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts:15]: all the types!

File: packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts [hunk 6]
```diff
@@ -51,10 +84,11 @@ export class OptimisticOracleClient {
     public readonly web3: Web3,
     oracleAddress: string,
     votingAddress: string,
-    public readonly lookback = 604800 // 1 Week
+    public readonly lookback: number = 604800, // 1 Week
+    public readonly oracleType: OptimisticOracleType = "OptimisticOracle"
```
Comment (chrismaree) [inline packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts:88]: you should do some validation on this type. check it is either `OptimisticOracle` or `SkinnyOptimisticOracle` or error

File: packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts (hunk 1 above)
Comment (chrismaree) [reply packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts:147]: nice.

File: packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts (hunk 6 above)
Comment (nicholaspai) [reply packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts:88]: Isn't this automatically validated by making `oracleType` have the `OptimisticOracleType` type? I'm not very familiar with typescript though so i could be very wrong

File: packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts [hunk 7]
```diff
@@ -138,39 +192,42 @@ export class OptimisticOracleClient {
...
+      .filter(isDefined);
 
     // Filter proposals based on their expiration timestamp:
-    const isExpired = (proposal: OptimisticOracleWeb3Events.ProposePrice["returnValues"]) => {
-      return Number(proposal.expirationTimestamp) <= Number(currentTime);
+    const isExpired = (proposal: ProposePriceReturnValues): boolean => {
+      if (this.oracleType === "SkinnyOptimisticOracle") {
+        return (
+          Number(
+            ((((proposal as unknown) as SkinnyProposePrice["returnValues"]).request as unknown) as SkinnyRequest)
```
Comment (chrismaree) [inline packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts:222]: nothing to change here but I'm not a fan of these `unknown` type casts all over the place like this. wish there was a cleaner pattern we could follow.

**Review by chrismaree — APPROVED** [review]
This is awesome! good work @nicholaspai

**Pull request merged**